proposals/0292-package-registry-service.md

```diff
@@ -0,0 +1,770 @@
+  for repositories with large histories.
+
+Many language ecosystems have a <dfn>package registry</dfn>, including
+[RubyGems] for Ruby,
+[PyPI] for Python,
+[npm] for JavaScript, and
+[crates.io] for Rust.
+In fact,
+many Swift developers develop apps today using
+[CocoaPods] and its index of libraries.
+
+A package registry can offer faster and more reliable dependency resolution
+than downloading dependencies using Git.
+It can also support other useful functionality,
+such as:
+
+* **Advisories**:
+  Vulnerabilities can be communicated directly to package consumers
+  in a timely manner.
+* **Discoverability**
+  Package maintainers can annotate their releases with project metadata,
+  including its authors, license, and other information.
+* **Search**
+  A registry can provide a standard interface for searching available packages,
+  or provide the information necessary for others to create a search index.
+* **Flexibility**
+  Swift Package Manager requires an external dependency to be
+  hosted in a Git repository with a package manifest located in its root,
+  which may be a barrier to adoption for some projects.
+  A package registry imposes no requirements on
+  version control software or project structure.
+
+## Proposed solution
+
+This proposal defines a standard interface for package registry services
+and describes how Swift Package Manager can integrate with them
+to download dependencies.
+
+The goal of this proposal is to make dependency resolution
+more available and reproducible.
+We believe our proposed solution
+can meet or exceed the current performance of dependency resolution
+and will allow for new functionality to be built in the future.
+
+## Detailed design
+
+### Package registry service
+
+A package registry service implements REST API endpoints
+for listing releases for a package,
+fetching information about a release,
+downloading the source archive for a release,
+and publishing a new release of a package.
+
+| Method | Path                                                 | Description                                      |
+| ------ | ---------------------------------------------------- | ------------------------------------------------ |
+| `GET`  | `/{package}`                                         | List package releases                            |
+| `GET`  | `/{package}/{version}`                               | Fetch metadata for a package release             |
+| `GET`  | `/{package}/{version}/Package.swift{?swift-version}` | Fetch manifest for a package release             |
+| `GET`  | `/{package}/{version}.zip`                           | Download source archive for a package release    |
+| `PUT`  | `/{package}/{version}{?commit,branch,tag,path,url}`  | Publish a package release or update its metadata |
+
+A formal specification for the package registry interface
+is provided alongside this proposal.
+In addition,
+an OpenAPI (v3) document and a reference implementation
```

Comment:
I've found the [reference implementation](https://github.com/mattt/swift-registry) but where is the OpenAPI (v3) document? I've also looked in the client reference implementation, and I and can't seem to find there either. Did I miss a link somewhere?

Reply:
@StanTwinB You can find the OpenAPI spec here: https://github.com/mattt/swift-package-registry-oas. Here's a link to some auto-generated docs for the specification: https://mattt.github.io/swift-package-registry-oas/